src/mlpack/methods/reinforcement_learning/replay/hindsight_replay.hpp

```diff
@@ -0,0 +1,423 @@
+        isEnd = iE;
+      }
+    }
+  }
+
+  /**
+   * Sample some experiences.
+   *
+   * @param sampledStates Sampled encoded states.
+   * @param sampledActions Sampled actions.
+   * @param sampledRewards Sampled rewards.
+   * @param sampledNextStates Sampled encoded next states.
+   * @param isTerminal Indicate whether corresponding next state is terminal
+   *        state.
+   */
+  void Sample(arma::mat& sampledStates,
+              std::vector<ActionType>& sampledActions,
+              arma::rowvec& sampledRewards,
+              arma::mat& sampledNextStates,
+              arma::irowvec& isTerminal)
+  {
+    size_t upperBound = full ? capacity : position;
+    arma::uvec sampledIndices = arma::randi<arma::uvec>(
+        batchSize, arma::distr_param(0, upperBound - 1));
+
+    sampledStates = states.cols(sampledIndices);
+    for (size_t t = 0; t < sampledIndices.n_rows; t ++)
+      sampledActions.push_back(actions[sampledIndices[t]]);
+    sampledRewards = rewards.elem(sampledIndices).t();
+    sampledNextStates = nextStates.cols(sampledIndices);
+    isTerminal = this->isTerminal.elem(sampledIndices).t();
+  }
+
+  /**
+   * Get the number of transitions in the memory.
+   *
+   * @return Actual used memory size.
+   */
+  const size_t& Size()
+  {
+    return full ? capacity : position;
+  }
+
+  /**
+   * Update the priorities of transitions and Update the gradients.
+   *
+   * @param target The learned value.
+   * @param sampledActions Agent's sampled action.
+   * @param nextActionValues Agent's next action.
+   * @param gradients The model's gradients.
+   */
+  void Update(arma::mat target,
+              std::vector<ActionType> sampledActions,
+              arma::mat nextActionValues,
+              arma::mat& gradients)
+  {
+    /* Do nothing for hindsight replay. */
+  }
+
+  //! Get the number of steps for n-step agent.
+  const size_t& NSteps() const { return nSteps; }
+
+ private:
+  //! Locally-stored number of examples of each sample.
+  size_t batchSize;
+
+  //! Locally-stored total memory limit.
+  size_t capacity;
+
+  //! Indicate the position to store new transition.
+  size_t position;
+
+  //! Locally-stored indicator that whether the memory is full or not.
+  bool full;
+
+  //! Locally-stored number of steps to look into the future.
+  size_t nSteps;
+
+  //! Locally-stored buffer containing n consecutive steps.
+  std::deque<Transition> nStepBuffer;
+
+  //! Locally-stored encoded previous states.
+  arma::mat states;
+
+  //! Locally-stored previous actions.
+  std::vector<ActionType> actions;
+
+  //! Locally-stored previous rewards.
+  arma::rowvec rewards;
+
+  //! Locally-stored encoded previous next states.
+  arma::mat nextStates;
+
+  //! Locally-stored termination information of previous experience.
+  arma::irowvec isTerminal;
+
+  //! Locally-stored encoded goal states.
+  arma::mat goals;
+
+  //! Locally-stored transitions of episode
+  std::vector<Transition> episodeTransitions;
```

Comment:
`arma::Col<Transition>` can have custom template too

Reply:
as far as i know we cant push_back in `arma::Col` right ?

Reply:
Ah that sucks, stick to vector then